src/Microsoft.AspNetCore.Routing/Matching/ILEmitTrieFactory.cs

```diff
@@ -0,0 +1,588 @@
+            var module = assembly.DefineDynamicModule(assemblyName, fileName);
+            var type = module.DefineType("ILEmitTrie");
+            var method = type.DefineMethod(
+                "GetDestination", 
+                MethodAttributes.Public | MethodAttributes.Static,
+                CallingConventions.Standard,
+                typeof(int),
+                new [] { typeof(string), typeof(int), typeof(int), };
+
+            GenerateMethodBody(method.GetILGenerator(), defaultDestination, exitDestination, entries, vectorize);
+
+            type.CreateTypeInfo();
+            assembly.Save(fileName);
+        }
+#endif
+
+        private class Locals
+        {
+            public Locals(ILGenerator il, bool vectorize)
+            {
+                P = il.DeclareLocal(typeof(byte).MakeByRefType());
+                Span = il.DeclareLocal(typeof(ReadOnlySpan<char>));
+
+                UInt16Value = il.DeclareLocal(typeof(ushort));
+
+                if (vectorize)
+                {
+                    UInt64Value = il.DeclareLocal(typeof(ulong));
+                    UInt64LowerIndicator = il.DeclareLocal(typeof(ulong));
+                    UInt64UpperIndicator = il.DeclareLocal(typeof(ulong));
+                }
+            }
+
+            /// <summary>
+            /// Holds current character when processing a character at a time.
+            /// </summary>
+            public LocalBuilder UInt16Value { get; set; }
+
+            /// <summary>
+            /// Holds current character when processing 4 characters at a time.
+            /// </summary>
+            public LocalBuilder UInt64Value { get; set; }
+
+            /// <summary>
+            /// Used to covert casing. See comments where it's used.
+            /// </summary>
+            public LocalBuilder UInt64LowerIndicator { get; set; }
+
+            /// <summary>
+            /// Used to covert casing. See comments where it's used.
+            /// </summary>
+            public LocalBuilder UInt64UpperIndicator { get; set; }
+
+            /// <summary>
+            /// Holds a 'ref byte' reference to the current character (in bytes).
+            /// </summary>
+            public LocalBuilder P { get; set; }
+
+            /// <summary>
+            /// Holds the relevant portion of the path as a Span[byte].
+            /// </summary>
+            public LocalBuilder Span { get; set; }
+        }
+
+        private class Labels
+        {
+            /// <summary>
+            /// Label to goto that will return the default destination (not a match).
+            /// </summary>
+            public Label ReturnDefault { get; set; }
+
+            /// <summary>
+            /// Label to goto that will return a sentinel value for non-ascii text.
+            /// </summary>
+            public Label ReturnNotAscii { get; set; }
+        }
+
+        private class Methods
+        {
+            public static readonly Methods Instance = new Methods();
```

Comment:
Add a comment explaining why there is a static instance instead of using a static class.

Reply:
I explained that this does caching. I didn't explain that particular tradeoff because it's not important. someone could change it in the future and I don't care 👍